<a href="https://colab.research.google.com/github/simulate111/Computer-vision2026ABO/blob/main/CompViz_part_I_nr_1_ipynb_txt.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# IT00CJ11 Computer Vision (5 cr. ECTS), spring 2025

**What is Computer Vision?**
![image.png](attachment:image.png)
(Source: Prof. F.-F. Li, Stanford Vision Lab.)

![image-2.png](attachment:image-2.png)
(Source: Prof. F.-F. Li, Stanford Vision Lab.)

![image-3.png](attachment:image-3.png)

**What would be good to know at this point?**
- Basics of programming, how to edit and run scripts, code structure and datatypes
- Some engineering mathematics, like how to work with vectors, gradients, matrices as well as elementary functions and derivatives

**Tools for the course:**
- A platform, chosen at your own convenience, from which you can run python code and jupyter notebooks
- The most important program libraries for python, such as numpy, matplotlib and OpenCV
- Willingness and patience to troubleshoot practical issues with these tools

**Basic Operations with OpenCV:**
1. Importing OpenCV
2. Loading an image
3. Image Shape/ Resolution
4. Displaying the image
5. Resizing the image
6. Converting to Grayscale

# **Image preprocessing**
**Image processing** is divided into analogue image processing and digital image processing.

**Digital image processing** is the use of computer algorithms to perform image processing on digital images. It allows a much wider range of algorithms to be applied to the input data - the aim of digital image processing is to improve the image data (features) by suppressing unwanted distortions and/or enhancement of some important image features so that our **AI models** can benefit from this improved data to work on.


**Are Black and White and Grayscale images the same?**

* Black and white (monochrome),has only two "colors", black (ink or toner) and white (no ink or toner). It is used for things like text, where you want everything that is a printed character to be black and the background white.

* Grayscale contains shades of grey and is used for reproducing images.

So most of the images you see are actually grayscale.

## Introduction

Let us play with an example image (pick out a color image of your own) to see how different standard operations in OpenCV work.  

I take the easy approach, i.e. the file is in the same folder as this Jupyter Notebook we're working on.

# Basic Operations With OpenCV
Let us look at various concepts ranging from loading images to resizing them and so on.

## Importing OpenCV
The first requirement is to import the OpenCV module.

In [5]:
import cv2

## Loading an image using OpenCV
We can read the image using imread module.
* It takes 2 arguments:
1. **path:** A string representing the path of the image to be read.
2. **flag:** It specifies the way in which image should be read. It’s default value is cv2.IMREAD_COLOR

Describing the 3 possible values of flag below:
* **cv2.IMREAD_COLOR:** It specifies to load a color image. Any transparency of image will be neglected. It is the default flag. Alternatively, we can pass integer value 1 for this flag.
* **cv2.IMREAD_GRAYSCALE:** It specifies to load an image in grayscale mode. Alternatively, we can pass integer value 0 for this flag.
* **cv2.IMREAD_UNCHANGED:** It specifies to load an image as such including alpha channel. Alternatively, we can pass integer value -1 for this flag.

In short:
* The 1 in the flag parameter denotes that it is a color image.
* If the parameter is 0, it would mean that the image being imported is a grayscale image.

Since the image I'm using is coloured, I'll use 1.

In [6]:
path = "roadmap.jpg"  # Add the path of the image here

In [7]:
# For colored Image
img = cv2.imread (path,1)   # Here, path can be either the path of image in your device, in Colab or a image URL

# For Black and White (gray scale)
# img = cv2.imread (path,0)

## Image Shape/Resolution etc.
We can make use of the shape sub-function to print out the shape of the image. Check out the below image:

In [8]:
img.shape

(845, 843, 3)

By shape of the image, we mean the shape of the NumPy array. As you see from executing the code, the matrix consists of 845 rows and 843 columns.

The last value 3 indicates the no. of channels.
* 3 = coloured image
* 1 = black and white/grayscale image



In [9]:
img.size

2137005

In [10]:
img.dtype

dtype('uint8')

## Working with image channels

Sometimes you will need to work separately on the B,G,R channels of an image. In this case, you need to split the BGR image into single channels. In other cases, you may need to join these individual channels to create a BGR image. You can do this simply by:

In [11]:
b,g,r = cv2.split(img)
print(r.shape)

rimg = cv2.merge([b*0, g*0, r*1])

# img[:,:,0] = 0
# img[:,:,1] = 0

(845, 843)


In [12]:
# For Jupyter Notebook users:
cv2.imshow("red",rimg)  # Here, "red" is the name that'll be given to the window

# waiting for key event
cv2.waitKey(0)

# destroying all windows
cv2.destroyAllWindows()

DisabledFunctionError: cv2.imshow() is disabled in Colab, because it causes Jupyter sessions
to crash; see https://github.com/jupyter/notebook/issues/3935.
As a substitution, consider using
  from google.colab.patches import cv2_imshow


## Displaying the image
Displaying an image using OpenCV is pretty simple and straightforward. Consider the below image:

**Note:** The imshow method of cv2 is disabled in Google Colab. Thus, colab users need to import cv2_imshow for displaying images.

So the commands will be like:
* Jupyter Notebook: cv2.imshow()
* Google Colab: cv2_imshow()

Now, just displaying the image often leads to crashing.

**The Problem**
The Notebook it is not usually happy to accommodate the window created by imshow. So it just crashes the window.

**The Solution**
To solve this, you need to explicitly close the window when you are done with it. To accomplish this, you can use the cv2.destroyAllWindows() functionality.

Since you probably don’t want your screen to close immediately, you can tell OpenCV to wait for a keypress.

* waitKey(0) will display the window infinitely until any keypress (it is suitable for image display).

* waitKey(1) will display a frame for 1 ms, after which display will be automatically closed. If you specify 25 inside the brackets, it will display for 25ms.



In [ ]:
# For Colab users
# from google.colab.patches import cv2_imshow
# cv2_imshow(img)

# For Jupyter Notebook users:
cv2.imshow("roadmp",img)  # Here, "roadmp" is the name that'll be given to the window

# waiting for key event
cv2.waitKey(0)

# destroying all windows
cv2.destroyAllWindows()

As you can see, we first import the image using imread. We require a window output to display the images, right?

We use the imshow function to display the image by opening a window.

Later, we wait for a user event. waitKey makes the window static until the user presses a key. The parameter passed to it is the time in milliseconds.

And lastly, we use destroyAllWindows to close the window based on the waitForKey parameter.

## Resizing the image:
Similarly, resizing an image is very easy. Here’s another code snippet:

In [ ]:
resized_image = cv2.resize(img, (500,300))

# Colab users:
# cv2_imshow(resized_image)

# Jupyter Notebook users:
cv2.imshow("roadmp", resized_image)

cv2.waitKey(0)
cv2.destroyAllWindows()

Here, resize function is used to resize an image to the desired shape. The parameter inside cv2.resize() is the shape of the new resized image.

Later, do note that the image object changes from img to resized_image, because of the image object changes now.

Rest of the code is pretty simple to the previous one, correct?




## Converting to Grayscale

cvtColor() function in OpenCV is very helpful in converting color channels from one to another such as BRG to HSV or BRG to RGB. The same method can be used to convert BRG to GRAY by using the cv2.cvtColor(img,cv2.BGR2GRAY)

Another method is to specify 0 instead of 1 as the flag parameter in imread(). That way, your image will be read as a grayscale image.

In [ ]:
# Converting color image to grayscale image
gray_img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

# Showing the converted image

# Colab users:
# cv2_imshow(gray_img)

# Jupyter Notebook users:
cv2.imshow("Converted Image",gray_img)

# waiting for key event
cv2.waitKey(0)

# destroying all windows
cv2.destroyAllWindows()